Support attributes in complex types in wsdl

jaydeepk opened the pull request:
**What**:
Implemented support for attributes in complex elements in WSDL

**Checklist**:

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [x] Tests
- [ ] Sonar Quality Gate

**Issue ID**:
Closes: #670

jaydeepk committed:
Implemented support for attributes in complex elements.

jaydeepk committed:
#670: Implemented support for optionality in attributes in wsdl.

jaydeepk committed:
Merge branch 'main' into support-attributes-in-complex-types-in-wsdl

joelrosario committed:
Minor refactorings

jaydeepk committed:
Removed explicit filtering of XMLPattern attributes to only retain the 'SPECMATIC_XML_ATTRIBUTE_PREFIX' attribute when de-referencing.
As due to this, any other custom attribute defined, gets ignored.

Added tests to cover scenarios with and without examples when:
1. A mandatory attribute is defined
2. An optional attribute is defined

jaydeepk committed:
Added tests to validate behaviour of mandatory attributes in wsdl in stub mode.

jaydeepk committed:
minor changes, clean up

jaydeepk committed:
Validated actual counts of expected tests instead of just asserting if the count is greater than 1.
Modified the stub tests for attributes to make them more lightweight and precise (without having to actually run the stub)

jaydeepk committed:
Created a method to log all the facts associated with a request, so that we can identify the different scenarios that are being tested ( for wsdl attribute tests)

jaydeepk committed:
Merge branch 'main' into support-attributes-in-complex-types-in-wsdl

joelrosario commented on core/src/test/kotlin/in/specmatic/conversions/WSDLConversionTests.kt:
```diff
@@ -473,6 +473,204 @@ class WSDLConversionTests {
+                    <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+                      <soapenv:Header specmatic_occurs="optional"/>
+                      <soapenv:Body>
+                        <Person age.opt="(number)" specmatic_type="SimpleOperation_SOAPPayload_Input"/>
```
Figuring out that the "age" attribute is optional took me a minute, even though we've paired, so it's probably harder for someone new to the test. Can you pull this name out as a variable named optionalAttributeName or something similar at the top of the test.

joelrosario commented on core/src/test/kotlin/in/specmatic/conversions/WSDLConversionTests.kt:
```diff
@@ -473,6 +473,204 @@ class WSDLConversionTests {
+                    <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+                      <soapenv:Header specmatic_occurs="optional"/>
+                      <soapenv:Body>
+                        <Person age="(number)" specmatic_type="SimpleOperation_SOAPPayload_Input"/>
```
Similar feedback to the test above.

joelrosario committed:
Updated some tests

- Changed how the specification is loaded in "should match soap request
  with attribute which is optional in the wsdl"
- Updated test names for test generation tests

joelrosario committed:
Updated some tests

- Changed how the specification is loaded in the new WSDL stub tests
- Updated test names for test generation tests

joelrosario committed:
Updated WSDL tests for attributes

- Renamed the tests
- Pulled out the attribute name into a variable so that it's easy to see
  the name of the attribute that's important to the tests

joelrosario reviewed (approved):
Looks good to me
